code/crates/test/app/src/state.rs

```diff
@@ -0,0 +1,385 @@
+        ))
+    }
+
+    /// Creates a new proposal value for the given height
+    /// Returns either a previously built proposal or creates a new one
+    pub async fn propose_value(
+        &mut self,
+        height: Height,
+        round: Round,
+    ) -> eyre::Result<LocallyProposedValue<TestContext>> {
+        assert_eq!(height, self.current_height);
+        assert_eq!(round, self.current_round);
+
+        let proposal = self.create_proposal(height, round).await?;
+
+        Ok(LocallyProposedValue::new(
+            proposal.height,
+            proposal.round,
+            proposal.value,
+        ))
+    }
+
+    /// Creates a stream message containing a proposal part.
+    /// Updates internal sequence number and current proposal.
+    pub fn stream_proposal(
+        &mut self,
+        value: LocallyProposedValue<TestContext>,
+    ) -> impl Iterator<Item = StreamMessage<ProposalPart>> {
+        let parts = self.value_to_parts(value);
+
+        let stream_id = self.stream_id;
+        self.stream_id += 1;
+
+        let mut msgs = Vec::with_capacity(parts.len() + 1);
+        let mut sequence = 0;
+
+        for part in parts {
+            let msg = StreamMessage::new(stream_id, sequence, StreamContent::Data(part));
+            sequence += 1;
+            msgs.push(msg);
+        }
+
+        msgs.push(StreamMessage::new(
+            stream_id,
+            sequence,
+            StreamContent::Fin(true),
+        ));
+
+        msgs.into_iter()
+    }
+
+    fn value_to_parts(&self, value: LocallyProposedValue<TestContext>) -> Vec<ProposalPart> {
+        let mut hasher = sha3::Keccak256::new();
+        let mut parts = Vec::new();
+
+        // Init
+        // Include metadata about the proposal
+        {
+            parts.push(ProposalPart::Init(ProposalInit::new(
+                value.height,
+                value.round,
+                self.address,
+            )));
+
+            hasher.update(value.height.as_u64().to_be_bytes().as_slice());
+            hasher.update(value.round.as_i64().to_be_bytes().as_slice());
+        }
+
+        // Data
+        // Include each prime factor of the value as a separate proposal part
+        {
+            for factor in factor_value(value.value) {
+                parts.push(ProposalPart::Data(ProposalData::new(factor)));
+
+                hasher.update(factor.to_be_bytes().as_slice());
+            }
+        }
+
+        // Fin
+        // Sign the hash of the proposal parts
+        {
+            let hash = hasher.finalize().to_vec();
+            let signature = self.signing_provider.sign(&hash);
+            parts.push(ProposalPart::Fin(ProposalFin::new(signature)));
+        }
+
+        parts
+    }
+}
+
+/// Re-assemble a [`ProposedValue`] from its [`ProposalParts`].
+///
+/// This is done by multiplying all the factors in the parts.
+fn assemble_value_from_parts(parts: ProposalParts) -> ProposedValue<TestContext> {
+    let value = parts
```

Comment:
Normally there should be a check that the received parts put together match some `ValueId` the proposer computed. I would put the `u64` in the Fin and double check it matches  `value` here. Alternatively, we should hash the parts and make sure they match the current hash in the Fin.
It's not super important for the use of this app in tests but rather if it's looked at as tutorial.

Reply:
We already check the signature before assembling the parts, I've refactored it slightly in 38843d80.

Reply:
Done in a6b77ec6